### PR TITLE
@impossible/large deposits withdraws

### PR DESCRIPTION
**Pull request by yasmine9**

### CODE REVIEW

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: Impossible-Agent-Suite/large-deposits-withdraws/src/large.sales.handler.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import { Finding, LogDescription, TransactionEvent } from "forta-agent";
+import SalesFetcher from "./sales.fetcher";
+import {
+  createSaleFinding,
+  PERCENT,
+  PURCHASE_SIGNATURE,
+  WITHDRAW_SIGNATURE,
+} from "./utils";
+
+export const provideSaleHandler = async (
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/large.sales.handler.ts]: changed to `saleHandler`

File: Impossible-Agent-Suite/large-deposits-withdraws/src/large.staking.handler.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import {
+  createStakeFinding,
+  IDIA_TOKEN,
+  PERCENT,
+  STAKE_SIGNATURE,
+  UNSTAKE_SIGNATURE,
+} from "./utils";
+import StakeFetcher from "./stake.fetcher";
+
+export const provideStakeHandler = async (
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/large.staking.handler.ts]: changed to `stakeHandler`

File: Impossible-Agent-Suite/large-deposits-withdraws/src/large.staking.handler.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  // Get stake/unstake events
+  const stakingEvents = txEvent.filterLog(
+    [STAKE_SIGNATURE, UNSTAKE_SIGNATURE],
+    staking_contract
+  );
+  if (stakingEvents.length == 0) return findings;
+  // compute totalSupply of the token
+  const totalSupply = await fetcher.getTotalSupply(
+    IDIA_TOKEN,
+    txEvent.blockNumber
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/large.staking.handler.ts]: addressed in the last commit

File: Impossible-Agent-Suite/large-deposits-withdraws/src/stake.fetcher.spec.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  beforeEach(() => mockProvider.clear());
+
+  it("should fetch the correct values", async () => {
+    initialize();
+
+    for (let [contract, block, supply] of TEST_DATA) {
+      const total: BigNumber = await fetcher.getTotalSupply(
+        contract,
+        block
+        //   trackId
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/stake.fetcher.spec.ts]: was using the argument in an older version, comment removed in the last commit

File: Impossible-Agent-Suite/large-deposits-withdraws/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    const tx1: TransactionEvent = new TestTransactionEvent()
+      .setBlock(532188)
+      .addAnonymousEventLog(staking_address, log1.data, ...log1.topics)
+      .addAnonymousEventLog(staking_address, log2.data, ...log2.topics);
+    const tx2: TransactionEvent = new TestTransactionEvent()
+      .addAnonymousEventLog(sale_addresses[0], log3.data, ...log3.topics)
+      .addAnonymousEventLog(sale_addresses[1], log4.data, ...log4.topics);
+
+    const handlerCalls = [handler(tx1), handler(tx2)];
+    const findings = await (await Promise.all(handlerCalls)).flat();
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/agent.spec.ts]: addressed in the last commit

File: Impossible-Agent-Suite/large-deposits-withdraws/src/sales.fetcher.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { providers, Contract, BigNumber } from "ethers";
+import LRU from "lru-cache";
+import { SALE_ABI } from "./utils";
+
+export default class SalesFetcher {
+  private provider: providers.Provider;
+  private cache: LRU<string, Promise<BigNumber | Set<string>>>;
+  private mock: boolean;
```
[the remark this answers is not in the document]
Comment (yasmine9) [reply Impossible-Agent-Suite/large-deposits-withdraws/src/sales.fetcher.ts]: undefined logic removed

**Review by e1Ru1o — APPROVED** [review]
Nice work

**Pull request merged**